src/metabase/models/setting.clj

```diff
@@ -1055,24 +1055,29 @@
 (defn- valid-trs-or-tru? [desc]
   (is-form? allowed-deferred-i18n-forms desc))
 
-(defn- validate-description-form
-  "Check that `description-form` is a i18n form (e.g. [[metabase.util.i18n/deferred-tru]]). Returns `description-form`
-  as-is."
+(defn- validate-description-form*
+  "Check that `description-form` is a i18n form (e.g. [[metabase.util.i18n/deferred-tru]]).
+   If not, return a form for an exception to throw at a later stage.
+   The reason for this strange behaviour is that we need to build the exception at compile time, but will only know
+   whether we should throw it once all the macro arguments have been evaluated at runtime."
   [description-form]
   (when-not (valid-trs-or-tru? description-form)
     ;; this doesn't need to be i18n'ed because it's a compile-time error.
-    (throw (ex-info (str "defsetting docstrings must be a *deferred* i18n form unless the Setting has"
-                         " `:visibilty` `:internal`, `:setter` `:none`, or is defined in a test namespace."
-                         (format " Got: ^%s %s"
-                                 (some-> description-form class (.getCanonicalName))
-                                 (pr-str description-form)))
-                    {:description-form description-form})))
-  description-form)
-
-(defn- in-test?
-  "Is `defsetting` currently being used in a test namespace?"
-  []
-  (str/ends-with? (ns-name *ns*) "-test"))
+    `(ex-info (str "defsetting docstrings must be a *deferred* i18n form unless the Setting has"
+                   " `:visibility` `:internal`, `:setter` `:none`, or is defined in a test namespace."
+                   (format " Got: ^%s %s"
+                           (some-> ~description-form class (.getCanonicalName))
+                           (pr-str ~description-form)))
+              {:description-form ~description-form})))
+
+;; This exists as its own method so that we can stub it in tests
+(defn- in-test? [ns] (str/ends-with? ns "-test"))
```

Comment:
```suggestion
(defn- ns-in-test? [ns] (str/ends-with? ns "-test"))
```

Reply:
2 things:
- I think we should prefix the name with `ns`, it was fine before if it does not take an argument
- we should call `ns-name` on `ns` because I usually think when we see `ns,` we think of it as a symbol, not str. So we should avoid that footgun.

Reply:
1. It feels non-idiomatic to repeat the name of the arg in the function name for a predicate, but I don't feel strongly.
2. The ns here *is* the symbol, it turns out the `str/end-with?` implicitly handles converting to a string.